src/main/java/duke/parser/Parser.java

```diff
@@ -0,0 +1,139 @@
+package duke.parser;
+
+import duke.commands.Deadline;
+import duke.commands.Event;
+import duke.commands.Task;
+import duke.commands.Todo;
+import duke.data.TaskList;
+import duke.exceptions.DukeException;
+import duke.ui.Ui;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.Arrays;
+
+/**
+ * Encapsulates a Parser object that reads in and executes commands based on user's input.
+ *
+ * @author Owen Tan
+ * @version Duke Level-9
+ */
+public class Parser {
+    private TaskList tasks;
+    private Ui ui;
+
+    /**
+     * Commands from user's input
+     */
+    public enum Keyword {
+        TODO, LIST, DEADLINE, EVENT, DONE, DELETE, FIND
+    }
+
+    /**
+     * Public constructor for Parser.
+     * @param tasks A list of tasks.
+     * @param ui Ui for Duke.
+     */
+    public Parser(TaskList tasks, Ui ui) {
+        this.tasks = tasks;
+        this.ui = ui;
+    }
+
+    /**
+     * Reads in and executes user's input based on string given.
+     * @param cmd String from user's input.
+     * @throws DukeException
+     */
+    public void parse(String cmd) throws DukeException {
+        String[] tokens = cmd.split(" ");
+        Keyword command = validateCommand(tokens);
+
+        int middle;
+        int index;
```

Comment:
Maybe can initialize variables here to void it being referred to while still being uninitialized?